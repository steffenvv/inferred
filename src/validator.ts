export interface PathContext {
    pushKey(key: string): void;
    popKey(): void;
    current(): string;
}

export interface ValidationContext {
    fail(message: string): never;
    readonly path: PathContext;
    typeName(value: any): string;
}

export interface Validator<T> {
    validate(value: any, context?: ValidationContext): T;
    isValid(value: any): value is T;
    readonly orNull: Validator<T | null>;
    readonly orUndefined: Validator<T | undefined>;
    readonly array: Validator<ReadonlyArray<T>>;
    or<U>(otherValidator: Validator<U>): Validator<T | U>;
}

export function keys<T>(obj: T): (keyof T)[] {
    return Object.keys(obj) as (keyof T)[];
}

function typeName(x: any): string {
    if (x === null) {
        return "null";
    } else if (x === undefined) {
        return "undefined";
    } else if (Array.isArray(x)) {
        return "an array";
    } else if (typeof x === "object") {
        return "an object";
    } else {
        return `a ${typeof x}`;
    }
}

export function makeContext(): ValidationContext {
    const path: string[] = [];

    const context: ValidationContext = {
        fail: (message: string): never => {
            const path = context.path.current();
            if (!path) {
                throw new Error(`Validation error: ${message}`);
            } else {
                throw new Error(`Validation error for key '${path}': ${message}`);
            }
        },

        path: {
            pushKey: key => path.push(key),

            popKey: () => path.pop(),

            current: () => path.join(".")
        },

        typeName
    };

    return context;
}

export function makeValidator<T>(validate: (value: any, context: ValidationContext) => T): Validator<T> {
    return {
        validate: (x, context = makeContext()) => {
            return validate(x, context);
        },

        isValid: (x: any): x is T => {
            try {
                validate(x, makeContext());
                return true;
            } catch {
                return false;
            }
        },

        get orNull(): Validator<T | null> {
            return makeValidator((x, context): T | null => {
                if (x === null) {
                    return null;
                }

                return validate(x, context);
            });
        },

        get orUndefined(): Validator<T | undefined> {
            return makeValidator((x, context): T | undefined => {
                if (x === undefined) {
                    return undefined;
                }

                return validate(x, context);
            });
        },

        get array(): Validator<ReadonlyArray<T>> {
            return makeValidator((x, context): ReadonlyArray<T> => {
                if (!Array.isArray(x)) {
                    return context.fail(`expected an array, not ${typeName(x)}`);
                }

                const result: T[] = [];
                let changed = false;

                for (let i = 0; i < x.length; i++) {
                    context.path.pushKey(i.toString());
                    try {
                        const value = x[i];
                        const validatedValue = validate(value, context);
                        result.push(validatedValue);
                        changed = changed || value !== validatedValue;
                    } finally {
                        context.path.popKey();
                    }
                }

                /* Preserve the object identity if nothing changed. */
                return changed ? result : x;
            });
        },

        or<U>(otherValidator: Validator<U>): Validator<T | U> {
            return makeValidator((x, context): T | U => {
                try {
                    return validate(x, context);
                } catch {
                    return otherValidator.validate(x, context);
                }
            });
        }
    };
}

export const aString = makeValidator(
    (x, context): string => (typeof x === "string" ? x : context.fail(`expected a string, not ${typeName(x)}`))
);

export const aBoolean = makeValidator(
    (x, context): boolean => (typeof x === "boolean" ? x : context.fail(`expected a boolean, not ${typeName(x)}`))
);

export const aNumber = makeValidator(
    (x, context): number => (typeof x === "number" ? x : context.fail(`expected a number, not ${typeName(x)}`))
);

export type ThunkValidator<T> = Validator<T> | (() => Validator<T>);

export type ValidatorMap<T> = { [K in keyof T]: ThunkValidator<T[K]> };

export type ExtendsUndefined<T> = T extends undefined ? true : false;

export type CanBeUndefined<T> = ExtendsUndefined<T> extends false ? false : true;

export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

export type OptionalKeys<T> = { [K in keyof T]: CanBeUndefined<T[K]> extends true ? K : never }[keyof T];

export type RequiredKeys<T> = Exclude<keyof T, OptionalKeys<T>>;

export type UndefinedToOptional<T> = OptionalKeys<T> extends never
    ? T
    : { [K in RequiredKeys<T>]: T[K] } & { [K in OptionalKeys<T>]?: T[K] | undefined };

export type Validated<T> = { readonly [K in keyof UndefinedToOptional<T>]: UndefinedToOptional<T>[K] };

export function anObject<T>(validators: ValidatorMap<T>): Validator<Validated<T>> {
    const validatorKeys = keys(validators);

    return makeValidator((x, context): Validated<T> => {
        if (x === null || typeof x !== "object") {
            return context.fail(`expected an object, not ${typeName(x)}`);
        }

        const result: any = {};
        let changed = false;

        for (const key of validatorKeys) {
            if (typeof key !== "string") {
                continue;
            }

            const value = x[key];
            const validator: ThunkValidator<any> = validators[key];
            const validate = typeof validator === "function" ? validator().validate : validator.validate;

            context.path.pushKey(key);
            try {
                const validatedValue = validate(value, context);
                if (validatedValue === undefined && !x.hasOwnProperty(key)) {
                    /* Don't introduce new keys for missing optional values. */
                } else {
                    result[key] = validatedValue;
                    changed = changed || validatedValue !== value;
                }
            } finally {
                context.path.popKey();
            }
        }

        /* Check for extra keys */
        for (const key of keys(x)) {
            if (typeof key === "string" && !validators.hasOwnProperty(key)) {
                return context.fail(`unexpected property "${key}"`);
            }
        }

        /* Preserve the object identity if nothing changed. */
        return changed ? result : x;
    });
}

export function aStringLiteral<T extends string>(value: T): Validator<T> {
    return aStringUnion(value);
}

export function aStringUnion<T extends string>(...values: T[]): Validator<T> {
    const expectedType = values.map(x => `"${x}"`).join(" | ");

    return makeValidator((x, context) => {
        if (values.indexOf(x) === -1) {
            return context.fail(`expected ${expectedType}, not ${typeName(x)}`);
        }

        return x;
    });
}

export type InferType<T extends Validator<any>> = T extends Validator<infer U> ? U : never;
