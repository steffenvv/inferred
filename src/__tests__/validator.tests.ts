import { Validator, aBoolean, aNumber, aString, anObject } from "../validator";

interface Item {
    name: string;
    value: string | null;
    enabled: boolean;
    maybeUndefined: string | undefined;
    subItems?: SubItem[];
    /* nextItem: Item | null; */
}

interface SubItem {
    name: string;
    value: number;
    meta?: string | null;
}

const itemValidator: Validator<Item> = anObject({
    name: aString,
    value: aString.orNull,
    enabled: aBoolean,
    maybeUndefined: aString.orUndefined,
    subItems: anObject({
        name: aString,
        value: aNumber,
        meta: aString.orNull.orUndefined
    }).array.orUndefined
});

const isValidItem = itemValidator.isValid;
const validateItem = itemValidator.validate;

describe("validation", () => {
    it("complains about missing properties", () => {
        expect(() => validateItem({})).toThrowErrorMatchingSnapshot();
        expect(() => validateItem({ name: "", enabled: false })).toThrowErrorMatchingSnapshot();
        expect(() => validateItem({ name: "", value: undefined, enabled: true })).toThrowErrorMatchingSnapshot();
        expect(() => validateItem({ name: "", value: null })).toThrowErrorMatchingSnapshot();
        expect(() =>
            validateItem({
                name: "name",
                value: null,
                enabled: false,
                subItems: [{ name: "", value: 42 }, { name: "" }]
            })
        ).toThrowErrorMatchingSnapshot();
    });

    it("does not complain about missing optional properties", () => {
        expect(validateItem({ name: "", value: null, enabled: false })).toMatchSnapshot();
    });

    it("does not insert undefined values for missing optional properties", () => {
        expect(validateItem({ name: "", value: null, enabled: true })).not.toHaveProperty("maybeUndefined");
        expect(validateItem({ name: "", value: null, enabled: true })).not.toHaveProperty("subItems");
    });

    it("removes extra properties when validating", () => {
        expect(validateItem({ name: "", value: null, enabled: false, foo: "bar" })).not.toHaveProperty("foo");
    });

    it("keeps undefined values when they are allowed", () => {
        expect(validateItem({ name: "", value: null, enabled: false, maybeUndefined: undefined })).toHaveProperty(
            "maybeUndefined"
        );
        expect(validateItem({ name: "", value: null, enabled: false, subItems: undefined })).toHaveProperty("subItems");
    });

    it("complains about values with the wrong type", () => {
        expect(() => validateItem({ name: 42, value: null, enabled: false })).toThrowErrorMatchingSnapshot();
        expect(() => validateItem({ name: "42", value: true, enabled: false })).toThrowErrorMatchingSnapshot();
        expect(() => validateItem({ name: "42", value: null, enabled: "false" })).toThrowErrorMatchingSnapshot();
        expect(() => validateItem({ name: "42", value: [null], enabled: false })).toThrowErrorMatchingSnapshot();
    });

    it("works as a type guard", () => {
        const test = (v: any) => {
            if (isValidItem(v)) {
                expect(typeof v.name).toBe("string");
                expect({ valid: true, ...v }).toMatchSnapshot();
            } else {
                expect({ valid: false, ...v }).toMatchSnapshot();
            }
        };

        test({});
        test({ name: 3 });
        test({ name: "a" });
        test({ name: "b", value: null });
        test({ name: "b", value: null, enabled: true });
    });
});
