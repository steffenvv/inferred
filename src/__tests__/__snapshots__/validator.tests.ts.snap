// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation complains about extra properties 1`] = `"Validation error: unexpected property \\"foo\\""`;

exports[`validation complains about extra properties 2`] = `"Validation error for key 'subItems.0': unexpected property \\"foo\\""`;

exports[`validation complains about missing properties 1`] = `"Validation error for key 'name': expected a string, not undefined"`;

exports[`validation complains about missing properties 2`] = `"Validation error for key 'value': expected a string, not undefined"`;

exports[`validation complains about missing properties 3`] = `"Validation error for key 'value': expected a string, not undefined"`;

exports[`validation complains about missing properties 4`] = `"Validation error for key 'enabled': expected a boolean, not undefined"`;

exports[`validation complains about missing properties 5`] = `"Validation error for key 'subItems.1.value': expected a string, not undefined"`;

exports[`validation complains about missing properties 6`] = `"Validation error for key 'nextItem.nextItem.enabled': expected a boolean, not a string"`;

exports[`validation complains about values with the wrong type 1`] = `"Validation error for key 'name': expected a string, not a number"`;

exports[`validation complains about values with the wrong type 2`] = `"Validation error for key 'value': expected a string, not a boolean"`;

exports[`validation complains about values with the wrong type 3`] = `"Validation error for key 'enabled': expected a boolean, not a string"`;

exports[`validation complains about values with the wrong type 4`] = `"Validation error for key 'value': expected a string, not an array"`;

exports[`validation complains about values with the wrong type 5`] = `"Validation error for key 'kind': expected \\"A\\" | \\"B\\" | \\"C\\", not a string"`;

exports[`validation complains about values with the wrong type 6`] = `"Validation error for key 'kind': expected \\"A\\" | \\"B\\" | \\"C\\", not a number"`;

exports[`validation does not complain about missing optional properties 1`] = `
Object {
  "enabled": false,
  "name": "",
  "value": null,
}
`;

exports[`validation works as a type guard 1`] = `
Object {
  "valid": false,
}
`;

exports[`validation works as a type guard 2`] = `
Object {
  "name": 3,
  "valid": false,
}
`;

exports[`validation works as a type guard 3`] = `
Object {
  "name": "a",
  "valid": false,
}
`;

exports[`validation works as a type guard 4`] = `
Object {
  "name": "b",
  "valid": false,
  "value": null,
}
`;

exports[`validation works as a type guard 5`] = `
Object {
  "enabled": true,
  "name": "b",
  "valid": true,
  "value": null,
}
`;

exports[`validation works as a type guard 6`] = `
Object {
  "enabled": true,
  "kind": "B",
  "name": "b",
  "valid": true,
  "value": null,
}
`;

exports[`validation works with union types 1`] = `"Validation error: expected an array, not a number"`;

exports[`validation works with union types 2`] = `"Validation error for key '0': expected a number, not an array"`;

exports[`validation works with union types 3`] = `"Validation error for key '0': expected a number, not a string"`;

exports[`validation works with union types 4`] = `"Validation error: expected an array, not null"`;

exports[`validation works with union types 5`] = `"Validation error: expected an array, not undefined"`;
